DREAM/Optimization/tsgGradientDescent.hpp

```diff
@@ -0,0 +1,154 @@
+ *
+ * UT-BATTELLE, LLC AND THE UNITED STATES GOVERNMENT MAKE NO REPRESENTATIONS AND DISCLAIM ALL WARRANTIES, BOTH EXPRESSED AND
+ * IMPLIED. THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_GRADIENT_DESCENT_HPP
+#define __TASMANIAN_GRADIENT_DESCENT_HPP
+
+#include "tsgOptimizationUtils.hpp"
+
+/*!
+ * \internal
+ * \file tsgGradientDescent.hpp
+ * \brief Gradient descent state and algorithm.
+ * \author Weiwei Kong & Miroslav Stoyanov
+ * \ingroup TasmanianOptimization
+ *
+ * Definition of the gradient descent state class and the particle swarm algorithm.
+ * \endinternal
+ */
+
+namespace TasOptimization {
+
+/*!
+ * \ingroup OptimizationState
+ * \brief Stores the information about a gradient descent run.
+ *
+ * \par class GradientDescentState
+ * An object of this class represents the components of a gradient descent run and all key parts of the swarm can be read through
+ * member methods. Methods are also provided to modify a subset of its parts.
+ *
+ * \par Constructors and Copy/Move assignment
+ * All constructors require information about an initial candidate point and initial stepsize. A method is available to retrieve
+ * the number of dimensions. No default constructor is available for this class, but copy/move constructors and assignment operator=
+ * overloads are available. The number of dimensions \b cannot be modified.
+ * - GradientDescentState()
+ * - getNumDimensions()
+ *
+ * \par Set and Get Candidate and Stepsize
+ * Throughout the run of gradient descent, the candidate point and stepsize may change. Methods are available to retrieve these
+ * data and set them manually.
+ * - getCandidate(), setCandidate()
+ * - getStepsize(), setStepsize()
+ *
+ * \par Gradient Descent Algorithm
+ * The gradient descent optimization algorithm can be used on a gradient descent algorithm with the goal of minimizing a particular
+ * objective function. The current candidate point, with respect to the given objective function, can be obtained from the state
+ * using the getCandidate() method.
+ * - GradientDescent()
+ *
+ * \par
+ * More information about the gradient descent algorithm can be found in the following paper:
+ *
+ * \par
+ * > Nesterov, Y. (2013). Gradient methods for minimizing composite functions. <em>Mathematical programming</em>, 140(1), 125-161.
+ *
+ */
+class GradientDescentState {
+  public:
+    //! \brief The default constructor is allowed.
```

Comment:
* documentation says `allow`
* the next line says `delete`
* if all else fails, we can make a meme out of this one

Reply:
On a related but tangent note, this is why "self-documented code" is a garbage notion.

Documentation gives you a layer of redundancy, if I see `GradientDescentState() = delete;` I know exactly what you are doing, but I have no way to check if this is what you wanted to do. When you have the code and the documentation, you double-down and affirm the correct thing (e.g., "yes, I want to delete this") or you make an error that we can now correct. Without documentation, an error in `default/delete` is much harder or catch.